### PR TITLE
Homework - make_unique

### PR DESCRIPTION
**Pull request by arafalski**

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
I dont have any comments 👍

**Review by WBigaj — APPROVED** [review]
I have checked your code and it looks good 👍

File: module1/homework/make_unique/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+add_executable(${PROJECT_NAME} main.cpp)
+target_compile_options(${PROJECT_NAME} PRIVATE ${FLAGS})
+target_link_options(${PROJECT_NAME} PRIVATE ${SANITS})
+
+add_executable(${PROJECT_NAME}-ut make_unique-ut.cpp)
+target_compile_options(${PROJECT_NAME}-ut PRIVATE ${FLAGS})
+target_link_libraries(${PROJECT_NAME}-ut PUBLIC gtest_main)
+target_link_options(${PROJECT_NAME}-ut PRIVATE ${SANITS})
+
+add_test(NAME shared_ptr COMMAND ${PROJECT_NAME}-ut)
```
Comment (ziobron) [inline module1/homework/make_unique/CMakeLists.txt]:
```suggestion
add_test(NAME make_unique COMMAND ${PROJECT_NAME}-ut)
```

File: module1/homework/make_unique/make_unique-ut.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    auto arr = cs::make_unique<int[]>(arraySize);
+
+    for (size_t i = 0; i < arraySize; ++i) {
+        arr[i] = value;
+    }
+
+    for (size_t i = 0; i < arraySize; ++i) {
+        ASSERT_EQ(arr[i], value);
+    }
+}
```
Comment (ziobron) [inline module1/homework/make_unique/make_unique-ut.cpp:93]: You have tests for passing various number of parameters 👍 
You have tests for arrays 👍 

You don't have tests for passing temporary values as parameters (by move). You need your own class on which you can verify the behavior - check if the move constructor / move assignment operator was called for r-values and copy constructor / copy assignment for l-values.

File: module1/homework/make_unique/make_unique.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#include <type_traits>
+
+namespace cs {
+
+template <typename T>
+using IsNotArrayType = std::enable_if_t<!std::is_array_v<T>>;
+
+template <typename T,
+          typename... Args,
+          typename = IsNotArrayType<T>>
```
Comment (ziobron) [inline module1/homework/make_unique/make_unique.hpp:13]: very nice and clean usage of SFINAE 👍

File: module1/homework/make_unique/make_unique.hpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+auto make_unique(Args&&... args) {
+    return std::unique_ptr<T>{new T{std::forward<Args>(args)...}};
+}
+
+template <typename T>
+auto make_unique(size_t size) {
+    return std::unique_ptr<T>{new std::remove_extent_t<T>[size] {}};
+}
```
Comment (ziobron) [inline module1/homework/make_unique/make_unique.hpp:21]: hmm... for me we can have possible memory leaks here. Can you check if valgrind reports something in MakeUniqueArrayTest? I can see that sanitizers are ok.

**Review by ziobron — CHANGES_REQUESTED** [review]
You don't have tests for passing temporary values as parameters (by move). You need your own class on which you can verify the behavior - check if the move constructor / move assignment operator was called for r-values and copy constructor / copy assignment for l-values.

File: module1/homework/make_unique/make_unique.hpp (hunk 4 above)
Comment (arafalski) [reply module1/homework/make_unique/make_unique.hpp:21]: Nope. Valgrind is clean
```
HEAP SUMMARY:
==23882==     in use at exit: 0 bytes in 0 blocks
==23882==   total heap usage: 277 allocs, 277 frees, 119,787 bytes allocated
==23882== 
==23882== All heap blocks were freed -- no leaks are possible
==23882== 
==23882== For lists of detected and suppressed errors, rerun with: -s
==23882== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Comment (ziobron) [reply module1/homework/make_unique/make_unique.hpp:21]: Yes, I thought that there will be T, but array is automatically deduced (eg. T is int[]), so there is no need to have T[]

**Review by ziobron — APPROVED** [review]
Nice, simple test for rvalue and lvalue 👍

**Comment by ziobron** [comment]
{ "requirement_ids": [1119, 4443], "points": [3, 40] }

**Pull request closed without merging**